bundles/org.eclipse.equinox.common/src/org/eclipse/core/runtime/ServiceCaller.java

```diff
@@ -160,6 +165,124 @@ public static <S> boolean callOnce(Class<?> caller, Class<S> serviceType, String
 		}).orElse(Boolean.FALSE);
 	}
 
+	/**
+	 * Not thread save.
+	 * <p>
+	 * Service registrations are obtained upon creation, service instances are get
+	 * on the first request (i.e. the first instance when {@link #first()} or
+	 * {@link #all()} is called. Services that are unregistered in the meantime are
+	 * discarded and newly registered services are ignored.
+	 * </p>
+	 * 
+	 * @param <S> the service type for this caller
+	 * @since 3.17
+	 */
+	public interface ServiceUsageScope<S> extends AutoCloseable { // TODO:? , Supplier<S>
+
+		/**
+		 * Gets the first available service implementation.
+		 */
+		Optional<S> first();
+
+		Stream<S> all();
```

Comment:
The javadoc on these methods need more precision. first should state it returns the highest ranked service, if there are any services. all should state the stream is ordered in ranking order (highest ranked service first, lowest ranked service last).

Reply:
Done.